File: mobile/ios/ShiftPlanner/ShiftPlanner/Shared/Models/Schedule.swift
import Foundation

enum AppScheduleStatus: String, Codable, Equatable {
    case draft
    case published
    case archived

    var title: String {
        switch self {
        case .draft:
            return localized("Draft", "Черновик")
        case .published:
            return localized("Published", "Опубликовано")
        case .archived:
            return localized("Archived", "Архив")
        }
    }
}

enum AppEmployeeAvailabilityStatus: String, Codable, Equatable {
    case available
    case ifNeeded = "if_needed"
    case unavailable

    /// Sort order: available first, then if-needed, then unavailable.
    var sortRank: Int {
        switch self {
        case .available: return 0
        case .ifNeeded: return 1
        case .unavailable: return 2
        }
    }

    var title: String {
        switch self {
        case .available:
            return localized("Available", "Доступен")
        case .ifNeeded:
            return localized("If needed", "При необходимости")
        case .unavailable:
            return localized("Unavailable", "Недоступен")
        }
    }

    static func fromApiValue(_ value: String) -> AppEmployeeAvailabilityStatus {
        AppEmployeeAvailabilityStatus(rawValue: value) ?? .unavailable
    }
}

struct AppScheduledShift: Identifiable, Equatable {
    let id: Int
    let employeeId: Int?
    let employeeName: String?
    let positionId: Int
    let positionName: String
    let date: Date
    let startMinutes: Int
    let endMinutes: Int

    var hasAssignedEmployee: Bool { employeeId != nil }
}

struct AppUnfilledRequirement: Identifiable, Equatable {
    let id: Int
    let positionId: Int
    let positionTitle: String
    let date: Date
    let startMinutes: Int
    let endMinutes: Int
    let missingStaff: Int
}

struct AppAvailableEmployee: Identifiable, Equatable {
    let id: Int
    let fullName: String
    let positionName: String
    let branchId: Int?
    let branchName: String?
    let availabilityStatus: AppEmployeeAvailabilityStatus
    let assignedHours: Double
}

/// Lightweight schedule descriptor from GET /schedule (list).
struct AppScheduleListItem: Identifiable, Equatable {
    let id: Int
    let branchId: Int?
    let startDate: Date
    let endDate: Date
    let status: AppScheduleStatus
}

/// Payload shared by create/update shift and requirement mutations.
struct ScheduleShiftMutation: Equatable {
    let date: Date
    let startMinutes: Int
    let endMinutes: Int
    let positionId: Int
    let employeeId: Int?
}

struct AppSchedule: Identifiable, Equatable {
    let id: Int
    let branchId: Int?
    /// Period covered by this schedule; nil when the backend didn't send it.
    let startDate: Date?
    let endDate: Date?
    let status: AppScheduleStatus
    let shifts: [AppScheduledShift]
    let unfilledRequirements: [AppUnfilledRequirement]

    var sortedShifts: [AppScheduledShift] {
        shifts.sorted {
            if $0.date == $1.date {
                if $0.startMinutes == $1.startMinutes {
                    return $0.positionName < $1.positionName
                }
                return $0.startMinutes < $1.startMinutes
            }
            return $0.date < $1.date
        }
    }

    var sortedUnfilledRequirements: [AppUnfilledRequirement] {
        unfilledRequirements.sorted {
            if $0.date == $1.date {
                if $0.startMinutes == $1.startMinutes {
                    return $0.positionTitle < $1.positionTitle
                }
                return $0.startMinutes < $1.startMinutes
            }
            return $0.date < $1.date
        }
    }

    var hasUnfilled: Bool {
        !unfilledRequirements.isEmpty || shifts.contains { !$0.hasAssignedEmployee }
    }
}

File: mobile/android/app/src/main/java/com/froggyriia/shiftplanner/domain/model/Schedule.kt
package com.froggyriia.shiftplanner.domain.model

import java.util.Date

enum class AppScheduleStatus(val apiValue: String) {
    DRAFT("draft"),
    PUBLISHED("published"),
    ARCHIVED("archived");

    companion object {
        fun fromApiValue(value: String): AppScheduleStatus? =
            entries.firstOrNull { it.apiValue == value }
    }
}

enum class AppEmployeeAvailabilityStatus(val apiValue: String) {
    AVAILABLE("available"),
    IF_NEEDED("if_needed"),
    UNAVAILABLE("unavailable");

    companion object {
        fun fromApiValue(value: String): AppEmployeeAvailabilityStatus =
            entries.firstOrNull { it.apiValue == value } ?: UNAVAILABLE
    }
}

data class AppAvailableEmployee(
    val id: Int,
    val fullName: String,
    val positionName: String,
    val branchId: Int? = null,
    val branchName: String?,
    val availabilityStatus: AppEmployeeAvailabilityStatus,
    val assignedHours: Double
)

data class AppScheduledShift(
    val id: Int,
    val employeeId: Int?,
    val employeeName: String?,
    val positionId: Int,
    val positionName: String,
    val date: Date,
    val startMinutes: Int,
    val endMinutes: Int
) {
    val hasAssignedEmployee: Boolean get() = employeeId != null
}

data class AppUnfilledRequirement(
    val id: Int,
    val positionId: Int,
    val positionTitle: String,
    val date: Date,
    val startMinutes: Int,
    val endMinutes: Int,
    val missingStaff: Int
)

data class AppSchedule(
    val id: Int,
    val branchId: Int?,
    /** Period covered by this schedule; null when the backend didn't send it. */
    val startDate: Date?,
    val endDate: Date?,
    val status: AppScheduleStatus,
    val shifts: List<AppScheduledShift>,
    val unfilledRequirements: List<AppUnfilledRequirement>
)

data class ScheduleShiftMutation(
    val date: Date,
    val startMinutes: Int,
    val endMinutes: Int,
    val positionId: Int,
    val employeeId: Int?
)

/** Lightweight schedule descriptor from GET /schedule (list). */
data class AppScheduleListItem(
    val id: Int,
    val branchId: Int?,
    val startDate: Date,
    val endDate: Date,
    val status: AppScheduleStatus
)

/** A shift-exchange request raised by an employee, reviewed by a manager. */
data class PendingShiftExchange(
    val id: Int,
    val shiftId: Int,
    val employeeId: Int,
    val employeeName: String,
    val note: String,
    val status: String   // "pending" | "approved" | "rejected"
)
